### PR TITLE
Post editor: iframe if all blocks are v3

### LINKED ISSUE #33346: Iframed Content Roadmap
**Goal**: to iframe all editor instances with front-end content.

It seems best to continue to **gradually** iframe the remaining content. For 5.9, I propose that we iframe the device previews and the block and pattern previews, and iframe the post editor at a later point. This is to give people time to test their blocks, address issues, contact us in case of problems, yet to slightly push them to do so. The previews (block, patterns and full content for device preview) are a good way to do that because they are non-essential while more visible.

## Done ✅

* [FSE](https://github.com/WordPress/gutenberg/pull/25775).
* [Template editor](https://github.com/WordPress/gutenberg/pull/31375).
* [Block, pattern and template previews](https://github.com/WordPress/gutenberg/pull/28165).
* [Device previews](https://github.com/WordPress/gutenberg/pull/33342).

## WP 5.9-6.0

* Prepare for the post editor.
  * Move the meta boxes (since the scroll container will move inside the iframe, while the meta boxes cannot). Would also allow us to remove the [Safari flickering hack](https://github.com/WordPress/gutenberg/pull/32824#issuecomment-869825472).
  * Backward compatibility layer (opt-in) to render blocks in an overlay outside the iframe if absolutely necessary.
    * This is necessary for the classic block, which currently doesn't work in the iframe.
* Refine the iframe component.
  * Remove event bubbling by rewriting:
    * All global shortcuts. Most of these how already been rewritten.
    * The dag chip #34819
  * Allow scripts and styles to be passed to editor settings in WP Dependency handle format, so themes and plugins can add to it. These handles should then be resolved on the server side into a list of assets, which could contain inline scripts and styles as well. See https://github.com/WordPress/gutenberg/issues/33212 and https://github.com/WordPress/gutenberg/issues/18571.
  * The assets that are loaded into the iframe need to much reduced. Currently we are loading way too many styles in the iframe that are meant for the UI. In some packages like `block-editor`, we'll need to create a separate stylesheet for editor content (that will be loaded in the iframe). We shouldn't load the whole package stylesheet which mostly consists of UI selectors.
* Move placeholders (editor UI) into shadow DOM. #33494
* Move appenders (and any other editor UI) out of the content and either into popovers or shadow DOM.
* Ideally the new widget and navigation screens should also be iframed. Who could best do that?

## WP 6.2

* ~[Post editor](https://github.com/WordPress/gutenberg/pull/21102)~ [Post editor](https://github.com/WordPress/gutenberg/pull/46212). Maybe this could be done earlier while it remains off by default. In some future release we could simply flip the switch. Another plugin could potentially flip the switch too. There was some interest for that.

## Dev notes

* [Blocks in an iframed (template) editor](https://make.wordpress.org/core/2021/06/29/blocks-in-an-iframed-template-editor/#comments).

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

This PR reintroduces the iframe for the post editor, but only if all registered blocks are v3 or higher.

Previously: #48076,  #46212, #21102.
Fixes #33346, #6156, #21193.

### CODE REVIEW

File: docs/reference-guides/block-api/block-metadata.md
Comment (fabiankaegy) [inline docs/reference-guides/block-api/block-metadata.md]: This here is still refering to version 2 being the latest one. We should add some context here to what V3 is all about / when it gets introduced.

**Review by fabiankaegy — COMMENTED** [review]
As always thank you for working on this :) I'm really excited about this! 

I know this is still in draft mode and you haven't asked for reviews yet. I added one comment where something is currently still missing but otherwise this is looking great from a code perspective :)

File: docs/reference-guides/block-api/block-metadata.md
Comment (ellatrix) [reply docs/reference-guides/block-api/block-metadata.md]: This is still a draft, I just batch replaced everything :)

**Comment by ellatrix** [comment]
Here's the plan: I'm going to force some e2e tests that I'm having a difficult time with to run without the iframe and address this later in follow-up PRs. I think it's important to get this PR in asap to get some testing before beta 1. It's just too overwhelming for me to fix the remaining e2e while continuously running into merge conflicts with trunk and people adding more e2e tests. The only thing this PR does is updating all blocks to v3 and fixing e2e tests.

File: packages/edit-post/src/components/visual-editor/index.js
Comment (mcsf) [inline packages/edit-post/src/components/visual-editor/index.js]: For clarity and performance, limit how much data is selected:

```suggestion
			hasV3BlocksOnly: getBlockTypes().every( ( type ) => {
 				return type.apiVersion >= 3;
		 	} ),
```

**Review by mcsf — COMMENTED** [review]
I have no objections to merging this, but I'll let others weigh in. Looking good. :)

File: packages/edit-post/src/components/visual-editor/index.js
Comment (ellatrix) [reply packages/edit-post/src/components/visual-editor/index.js]: I don't think we can do this because you'd return a new array on every call?

Comment (ellatrix) [reply packages/edit-post/src/components/visual-editor/index.js]: Ah no it's boolean. I don't remember why I did this then.

Comment (mcsf) [reply packages/edit-post/src/components/visual-editor/index.js]: Well, with `.every` you'd return a boolean, which is why it's nice. Right?

That said, realistically `getBlockTypes` rarely ever changes, so ignore my comment. 🙃  I guess an alternative take would be

```js
const hasV3BlocksOnly = useMemo( () => blockTypes.every( ... ), [ blockTypes ] )
```

but I doubt that it matters for a ~100 collection like block types.

**Comment by oandregal** [comment]
I've done some testing:

- :green_circle:  When I activate a plugin block that has api < 3, the iframe is not activated.
- :green_circle: TwentyTwentyOne: reviewed the "Gutenberg demo" post with and without iframe: it looks the same.
- :green_circle: : TwentyTwentyOne: added a custom background color via the "Customizer > Colors & Dark mode". The iframed editor doesn't pick it up. The theme uses `enqueue_block_editor_assets` for providing custom styles ([see](https://github.com/WordPress/wordpress-develop/blob/413f59caeea4a29b8b8a89d8970c9e3bd2c8578b/src/wp-content/themes/twentytwentyone/classes/class-twenty-twenty-one-custom-colors.php#L99)). Updated: fixed by https://github.com/WordPress/gutenberg/pull/48286/commits/85329a69cc5ae9fcc54d2cf11e86b5a6d87ef3eb
- :green_circle: : TwentyTwenty: reviewed the "Gutenberg demo" post with and without the iframe. I found spacing is different from `trunk` for lists (edit: fixed at https://github.com/WordPress/gutenberg/pull/48286/commits/898f62c28f60e2a5ec78a8eefed4db2432c834a6)

| Without iframe | With iframe |
| --- | --- |
| Captura de ecrã de 2023-06-02 12-55-11 | Captura de ecrã de 2023-06-02 12-55-18 |

File: packages/edit-post/src/components/visual-editor/index.js
Comment (ellatrix) [reply packages/edit-post/src/components/visual-editor/index.js]: Fixed in https://github.com/WordPress/gutenberg/pull/48286/commits/8423f85ab99c83ac43a71b545ac64f68301812cf. :)

File: packages/block-editor/src/components/iframe/use-compatibility-styles.js
Comment (ellatrix) [inline packages/block-editor/src/components/iframe/use-compatibility-styles.js:80]: We no longer need this comment because we added the warning in https://github.com/WordPress/gutenberg/pull/50091.

**Comment by oandregal** [comment]
I've also tested TT2 and TT3 with and without style modifications: they work as expected.

**Review by oandregal — APPROVED** [review]

**Comment by ellatrix** [comment]
Thank you for the review!

**Comment by fabiankaegy** [comment]
@bph Just raising this for you :) This will definitely need a Dev note when it gets merged into core :)

**Comment by oandregal** [comment]
Follow-up https://github.com/WordPress/gutenberg/pull/51199

**Comment by davidwebca** [comment]
Congrats on the great work leading this! @ellatrix 🤘

**Comment by bph** [comment]
> Just raising this for you :) This will definitely need a Dev note when it gets merged into core :) 

Thanks, @fabiankaegy  -  @Leonardus-Nugraha is your 6.3 Dev Notes wrangler for GB PRs. So keep flagging things you find, so he has it a little easier doing it for the first time.

File: test/e2e/specs/editor/various/keep-styles-on-block-transforms.spec.js
Comment (Mamaduka) [inline test/e2e/specs/editor/various/keep-styles-on-block-transforms.spec.js:46]: @ellatrix, what was the reason for adding this hack?

File: lib/compat/wordpress-6.3/script-loader.php
Comment (ramonjd) [inline lib/compat/wordpress-6.3/script-loader.php:50]: @ellatrix  

Does this need to be backported to WordPress 6.3 in [wp_includes/block-editor.php](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/block-editor.php#L341)?

**Comment by webmandesign** [comment]
Hi. Is there any way I can apply custom class on `.editor-styles-wrapper` wrapper [now that it is within an iframe](https://github.com/WordPress/gutenberg/issues/17854#issuecomment-1638268100)?

**Comment by markhowellsmead** [comment]
Until now, we've used `enqueue_block_editor_assets` to enqueue non-block-based scripts in the editors. Because `enqueue_block_assets` is apparently also run in the frontend, then do we need to do a manual check now, to see whether we're in the frontend or not?

**Comment by paulgibbs** [comment]
@ellatrix I share @markhowellsmead 's question above. Is this the correct approach we should have been taking, and if so, how can we find out if the action is run on the back or front of the site? Does `is_admin()` work still?

**Comment by ellatrix** [comment]
The problem with `enqueue_block_editor_assets` is that is has always been to style editor UI, it's not supposed to be used for editor content. You can indeed use `enqueue_block_assets` with an `is_admin()` check. If there's a lot of use cases, we could add another action for the editor content only.

**Comment by paulgibbs** [comment]
Thank you for the prompt reply and all of your hard work and contributions to WordPress, @ellatrix 👍🏻  🥇

**Comment by markhowellsmead** [comment]
Thanks from me too @ellatrix ! Which hook should we be using for editor scripts then? e.g. for enqueuing scripts which apply `addFilter` across multiple blocks, for registering custom data stores, or for adding plugin settings panels?

**Comment by ndiego** [comment]
> Which hook should we be using for editor scripts then? e.g. for enqueuing scripts which apply addFilter across multiple blocks, for registering custom data stores, or for adding plugin settings panels?

@markhowellsmead I would still use `enqueue_block_editor_assets`, but @ellatrix can confirm. I am using this in all my plugins and am not seeing any issues in 6.3.

**Comment by ellatrix** [comment]
> Which hook should we be using for editor scripts then? e.g. for enqueuing scripts which apply `addFilter` across multiple blocks, for registering custom data stores, or for adding plugin settings panels?

That's correct, for editor UI scripts, you should use `enqueue_block_editor_assets`. :)

**Comment by ellatrix** [comment]
I added some extra info to the iframe dev note: https://make.wordpress.org/core/2023/07/18/miscellaneous-editor-changes-in-wordpress-6-3/#post-editor-iframed

**Comment by markhowellsmead** [comment]
That’s great, thanks @ellatrix!

**Comment by webmandesign** [comment]
@ellatrix 
> The problem with enqueue_block_editor_assets is that is has always been to style editor UI, it's not supposed to be used for editor content. You can indeed use enqueue_block_assets with an is_admin() check. If there's a lot of use cases, we could add another action for the editor content only.

I'm quite confused here...

I was using the `enqueue_block_editor_assets` hook previously to add any block editor related assets, including CSS for my custom [block styles](https://developer.wordpress.org/reference/functions/register_block_style/). It was working perfectly fine.

After updating to WP6.3 and changing `enqueue_block_editor_assets` hook for `enqueue_block_assets` hook (when not dealing with editor UI), the CSS for block styles no longer works.
So I switched back to `enqueue_block_editor_assets` hook and it works again. Is CSS related to block style considered a UI related asset?

Can you please explain when exactly should we use `enqueue_block_editor_assets` hook and `enqueue_block_assets` hook? Thanks.

**Comment by webmandesign** [comment]
And can I keep using `enqueue_block_editor_assets` hook, is it future-proof?

**Comment by ellatrix** [comment]
For blocks, you should use block.json. I'm not sure how exactly `register_block_style`, but styles added through that should be automatically added to the iframe without needing to do anything else.

When it comes to other styles, `enqueue_block_editor_assets` should be used if you're styling inspector controls for example (UI), and `enqueue_block_assets` should be used when styling editor content.

**Comment by ndiego** [comment]
@ellatrix it has been fairly common practice (i.e. we have used it in a lot of [examples](https://developer.wordpress.org/news/2023/02/creating-custom-block-styles-in-wordpress-themes/)), to register/unregister block styles and variations in Javascript and then enqueue the file with `enqueue_block_editor_assets`. Something like this. 

```js
// assets/js/block-editor.js

wp.domReady( () => {
    wp.blocks.registerBlockStyle( 'core/navigation-link', {
        name: 'navigation-button',
        label: 'Button'
    } );
} );
```

```php
// functions.php

add_action( 'enqueue_block_editor_assets', 'themeslug_block_editor_assets' );

function themeslug_block_editor_assets() {
    wp_enqueue_script(
        'themeslug-block-editor',
        get_theme_file_uri( 'assets/js/block-editor.js' ),
        array( 
            'wp-blocks', 
            'wp-dom-ready', 
            'wp-edit-post' 
        )
    );
}
```

@webmandesign if you want to register/unregister block styles and variations, this is the approach I would take unless @ellatrix thinks differently. Of course, you can also use `register_block_style` for block styles. 

When adding your block-style CSS, I would use [`wp_enqueue_block_style()`](https://developer.wordpress.org/reference/functions/wp_enqueue_block_style/). @justintadlock has written a good article on this as well as another great one on leveraging per-block stylesheets:

- [Creating custom block styles in WordPress themes](https://developer.wordpress.org/news/2023/02/creating-custom-block-styles-in-wordpress-themes/)
- [Leveraging theme.json and per-block styles for more performant themes](https://developer.wordpress.org/news/2022/12/leveraging-theme-json-and-per-block-styles-for-more-performant-themes/)

**Comment by ellatrix** [comment]
`wp.blocks.registerBlockStyle` should definitely be called from `enqueue_block_editor_assets`. You will almost never want to add scripts through `enqueue_block_assets`, unless they're some sort of view script for the front-end as well. `registerBlockStyle` is an editor API, so it must be enqueued through `enqueue_block_editor_assets`.

**Comment by ndiego** [comment]
Perfect, thanks for the clarification @ellatrix 🙏

**Comment by markhowellsmead** [comment]
@ndiego I think this conflicts with the information [you posted](https://make.wordpress.org/core/2023/07/18/miscellaneous-editor-changes-in-wordpress-6-3/#post-editor-iframed) on the Make blog, doesn't it?

> To make adoption easier, all assets (styles and scripts) added through the enqueue_block_assets PHP action will now also be enqueued for the iframe.

Addition of styles is clear; addition of block registration and block-specific code is also clear. But what about sidebar panels, data store registration and so on?

**Comment by ellatrix** [comment]
> Addition of styles is clear; addition of block registration and block-specific code is also clear. But what about sidebar panels, data store registration and so on?

What do you mean? This is all editor UI, and it's all editor APIs. `enqueue_block_assets` is almost only for content styles with some rare exceptions for view scripts that will also load on the front end.

**Comment by markhowellsmead** [comment]
Thanks for continuing to help clarify these new changes. I have no questions related to styles: I'm asking about non-block-related scripts which need to load in the editors.

Unless I'm reading it incorrectly, the dev note on the [Make blog](https://make.wordpress.org/core/2023/07/18/miscellaneous-editor-changes-in-wordpress-6-3/#post-editor-iframed), which only really details how to approach styles, says that developers should be using `enqueue_block_assets` for scripts and styles. However, if “_`enqueue_block_assets` is almost only for content styles_”  then it sounds as though we shouldn't use it to enqueue scripts. Is that correct?

Your comment (https://github.com/WordPress/gutenberg/pull/48286#issuecomment-1661882101) states “_`enqueue_block_editor_assets` […] has always been to style editor UI_”: that means we shouldn't use it to load scripts in the editors. Is that correct?

If both of those are correct, then which hook do we use to load non-block scripts in the framed editors?

**Comment by justintadlock** [comment]
If you need scripts/styles for the editor UI, use `enqueue_block_editor_assets`.

If you need scripts/styles for the editor content (the part in the `<iframe>`) and front-end view, use `enqueue_block_assets`. And in the rare case where you need an admin-only script/style for the editor content but not the front-end, use an `is_admin()` check.

**Comment by ellatrix** [comment]
Yes. I understand it's a bit confusing that you have to use `enqueue_block_editor_assets` for stuff like `registerBlockType`, arguably a block is content, but those are all editor APIs, these are not on the front-end or in the iframe. `registerBlockType` etc. also do a bunch of other stuff like adding controls.

**Comment by markhowellsmead** [comment]
That's great, thanks Justin and Ella.

**Comment by webmandesign** [comment]
@ellatrix @ndiego @justintadlock 

Sorry for my late answer here. After some investigation and debugging I was able to trace the source of the issue to my code.
So, everything is working just like you say. Sorry for confusion and thank you for explanation.

(During debugging I've come across an issue of identifying styles, so I've logged a suggection at #53756)

**Comment by briangrider** [comment]
Hi @ellatrix,

Thank you for all of your hard work on this. The consistency of the iframed editor being available across different contexts is a huge step up in what can be done by 3rd party devs moving forward. I noticed that whenever custom fields are enabled in the post editor, the editor is no longer iframed. This is causing some issues with a plugin I'm working on. Is there a reason for this or is this just something that hasn't been looked at yet?

**Comment by ndiego** [comment]
>  I noticed that whenever custom fields are enabled in the post editor, the editor is no longer iframed. 

Yeah, this is intentional for backward compatibility. Whenever metaboxes are present, the Editor is also no longer iframed.

**Comment by briangrider** [comment]
Thanks @ndiego! Editing my post here because I realized I didn't ask questions I meant to ask. Do potential issues with backward compatibility have to do with the metabox section being in the iframe as well? What could potentially break if the editor is iframed but the metabox section is left outside of the iframe?

@ellatrix & @ndiego - does this mean that there is no immediate plan to move toward an iframed editor in all editor contexts, including when metaboxes are present?

Just my opinion, but with almost every editor context iframed now, it feels a bit unintuitive for that singular context to be un-iframed. Also, since posts/pages can easily be very long with the block editor, having custom fields that far out of reach seems unintuitive as well. I'm not really sure if this is the right place for posting ideas (and please let me know if there's a better way to address things like this), but here are some ideas I have:

Idea 1) When enabled, the custom fields section could be a separate { position: sticky; bottom: 0; } element outside of the editor iframe. Similar to how it is now except it would always be stuck to the bottom of the .interface-interface-skeleton__content container. It actually seems like the structure is already there for this when the editor is iframed (ie. metaboxes are disabled), but the metabox data just isn't populated so the dom structure content is blank. This is the current dom structure when the editor is iframed:

metaboxes-structure

If this is possible without unacceptably breaking backward compatibility, it would make custom fields more accessible while still not taking up too much space in the editor when collapsed. It also leaves the door open for eventually adding a show custom fields option to the site editor which might be desirable at some point. I haven't checked how the custom fields section appears on mobile but I assume it would be more manageable as a separate section in that context as well.

Idea 2) Add a togglable modal that contains the custom field section. This would most likely be located in the post sidebar. Ideally, custom fields make the most semantic sense there anyway (if the section wasn't so large).

If either of these are possible without breaking backward compatibility, I think idea 1 is better as it would feel familiar, and would allow for an iframed editor and easier access to custom fields for users.

Let me know your thoughts!

**Comment by ndiego** [comment]
I can't speak for @ellatrix but backward compatibility is the reason things are moving so slowly here. There are millions of WordPress users who run sites that use metaboxes, often with sophisticated implementations that were originally designed for a "Classic" environment. 

Custom fields is Core functionality, so I could see that being migrated sooner rather than later, but that does not solve third-party metaboxes. As I am sure Ella can attest, this is an extremely tricky thing to solve without breaking things. 

From a Developer Relations perspective, the goal now is to help/encourage plugin authors to migrate away from metaboxes and update all custom blocks to API v3.

**Comment by ellatrix** [comment]
@briangrider See https://github.com/WordPress/gutenberg/pull/34421 for a previous attempt.

Actually there's no backward compatibility issues with metaboxes and the iframe. The metaboxes must be rendered outside the iframe of course. The problem is that there's now two scrollable areas that need to fit into that space, which is a no-go. Alternatively we need there to be some sort of accordion to switch between content and meta boxes, or they need to be moved to a modal (that would be my preference since we're talking about a legacy API).

If you have idea, I'm happy to review any PR

**Comment by briangrider** [comment]
Thank you for your response @ellatrix. For your previous attempt, was it just an issue of having more important things to take care of that made you put it on hold? I've never contributed to Core but would love to eventually. I can't jump on this right now but if it hasn't been addressed in a few months, I'll take a look at the code and see if I can add anything valuable to the conversation.

**Comment by abredikis** [comment]
Given that there are no compatibility issues with metaboxes and iframes, could we explore the option of allowing to opt-in for iframe functionality even when metaboxes are active and maybe enabling the tab feature from [#34421](https://github.com/WordPress/gutenberg/pull/34421)? Otherwise the only way to use the iframe editor is to not use plugins that rely on metaboxes, which in my opinion is way worse than the alternative.

**Comment by giannis-koulouris-bb** [comment]
Agree with @abredikis, we need some way to enable the iframed editor even with metaboxes present, just had this issue with a website I'm building, used v3 on all blocks to make sure we get an iframed editor, and everything worked just fine, until we had to install an SEO plugin, but doing so disables the iframed editor because they make use of metaboxes.

**Comment by ellatrix** [comment]
See #59242 for the latest on meta boxes.

**Pull request merged**
